src/db/mysql/models.rs

```diff
@@ -90,16 +99,80 @@ impl MysqlDb {
         }
     }
 
-    pub(super) fn create_collection(&self, name: &str) -> Result<i32> {
-        // XXX: handle concurrent attempts at inserts
-        let id = self.conn.transaction(|| {
-            sql_query("INSERT INTO collections (name) VALUES (?)")
-                .bind::<Text, _>(name)
-                .execute(&self.conn)?;
-            collections::table.select(last_insert_id).first(&self.conn)
-        })?;
-        self.coll_cache.put(id, name.to_owned())?;
-        Ok(id)
+    /// APIs for collection-level locking
+    ///
+    /// Explicitly lock the matching row in the user_collections table. Read
+    /// locks do SELECT ... LOCK IN SHARE MODE and write locks do SELECT
+    /// ... FOR UPDATE.
+    ///
+    /// In theory it would be possible to use serializable transactions rather
+    /// than explicit locking, but our ops team have expressed concerns about
+    /// the efficiency of that approach at scale.
+    pub fn lock_for_read(&mut self, user_id: HawkIdentifier, collection: &str) -> Result<()> {
+        let user_id = user_id.legacy_id as u32;
+        let collection_id = self
+            .get_collection_id(collection)
+            .or_else(|e| match e.kind() {
+                // If the collection doesn't exist, we still want to start a
+                // transaction so it will continue to not exist.
+                DbErrorKind::CollectionNotFound => Ok(0),
+                _ => Err(e),
+            })?;
+        // If we already have a read or write lock then it's safe to
+        // use it as-is.
+        if let Some(_) = self.session.coll_locks.get(&(user_id, collection_id)) {
+            return Ok(());
+        }
+
+        // XXX: begin transaction here like python?
```

Comment:
If this is called outside of a transaction, it won't do anything useful, so I think we do need to resolve transaction-handling as part of this.  Would it make sense for the `MysqlDbSession` to "own" a transaction that is started when creating the session, and committed (or rolled back) when the session is dropped?  I'm not sure how that plays with ownership throughout the rest of the code here...